crypto/aes/asm/aes-s390x.pl

```diff
@@ -2219,6 +2254,69 @@ ()
 .size	AES_xts_decrypt,.-AES_xts_decrypt
 ___
 }
+
+################
+# void s390x_aes_gcm_blocks(unsigned char *out, GCM128_CONTEXT *ctx,
+#                           const unsigned char *in, size_t len,
+#                           const unsigned char *aad, size_t alen,
+#                           const AES_KEY *key, int enc)
+{
+my ($out,$ctx,$in,$len,$aad,$alen,$key,$enc) = map("%r$_",(2..9));
+$code.=<<___;
+.globl	s390x_aes_gcm_blocks
+.type	s390x_aes_gcm_blocks,\@function
+.align	16
+s390x_aes_gcm_blocks:
+.cfi_startproc
+	stm$g	$alen,$enc,7*$SIZE_T($sp)
+.cfi_rel_offset $alen,7*$SIZE_T
+.cfi_rel_offset $key,8*$SIZE_T
+.cfi_rel_offset $enc,9*$SIZE_T
+	lm$g	$alen,$enc,$stdframe($sp)
+
+	aghi	$sp,-112
+.cfi_adjust_cfa_offset 112
+
+	lmg	%r0,%r1,0($ctx)
+	ahi	%r1,-1			# kma requires counter-1
+
+	mvc	16(32,$sp),64($ctx)	# copy Xi/H
+	#mvc	48(16,$sp),48($ctx)	# copy len
+	mvc	80(32,$sp),0($key)	# copy key
+	st	%r1,12($sp)		# copy Yi
```

Comment:
One can wonder if it would be appropriate to use Htable as descriptor. Right now Htable is initialized for software processing with rationale that single-block [or just multiplication] is faster in software. But if it's not true anymore, or difference is not that big, then it might make sense... Well, I haven't actually examines C glue that closely yet, which means that this remark might be totally off. In other words, don't do anything about this yet...

Reply:
The reason not using the gcm contect to store the kma param block was the following:
The problem is see is with partial block processing during update calls, where i need to do a "partial encryption" (i.e. using the first bytes of the aes-ctr keystream and saving the remaining bytes for the next update call), while at the same time, authentication has to wait for the next update call. This just cannot be done using kma, and i have to rely on the old code path in those cases.